builtin/common/after.lua

```diff
@@ -9,42 +107,55 @@ core.register_globalstep(function(dtime)
 		return
 	end
 
-	time_next = math.huge
-
-	-- Iterate backwards so that we miss any new timers added by
-	-- a timer callback.
-	for i = #jobs, 1, -1 do
-		local job = jobs[i]
-		if time >= job.expire then
-			core.set_last_run_mod(job.mod_origin)
-			job.func(unpack(job.arg))
-			local jobs_l = #jobs
-			jobs[i] = jobs[jobs_l]
-			jobs[jobs_l] = nil
-		elseif job.expire < time_next then
-			time_next = job.expire
+	-- Remove the expired jobs.
+	local expired = remove_first_jobs()
+
+	-- Remove other expired jobs and append them to the list.
+	while true do
+		time_next = expiries[1] or math.huge
+		if time_next > time then
+			break
+		end
+		list_append_list(expired, remove_first_jobs())
+	end
+
+	-- Run the callbacks afterward to prevent infinite loops with core.after(0, ...).
+	local last_expired = expired.list_end
+	while true do
+		core.set_last_run_mod(expired.mod_origin)
+		expired.func(unpack(expired, 1, expired.n_args))
+		if expired == last_expired then
+			break
 		end
+		expired = expired.list_next
 	end
 end)
 
+local job_metatable = {__index = {}}
+
+local function dummy_func() end
+function job_metatable.__index:cancel()
+	self.func = dummy_func
+	for i = 1, self.n_args do
+		self[i] = nil
+	end
+	self.n_args = 0
+end
+
 function core.after(after, func, ...)
-	assert(tonumber(after) and type(func) == "function",
+	assert(tonumber(after) and not core.is_nan(after) and type(func) == "function",
 		"Invalid minetest.after invocation")
-	local expire = time + after
+
 	local new_job = {
-		func = func,
-		expire = expire,
-		arg = {...},
 		mod_origin = core.get_last_run_mod(),
+		func = func,
+		n_args = select("#", ...),
+		...
 	}
 
-	jobs[#jobs + 1] = new_job
-	time_next = math.min(time_next, expire)
+	local expiry = time + after
+	add_job(expiry, new_job)
+	time_next = math.min(time_next, expiry)
 
-	return {
-		cancel = function()
-			new_job.func = function() end
-			new_job.args = {}
-		end
-	}
+	return setmetatable(new_job, job_metatable)
```

Comment:
wait this doesn't mean mods get a handle to the internal job table, do they?

Reply:
They get the job object (as before) and can obtain the job metatable (where cancel is implemented) via `getmetatable`.

If this is a problem, we could use `__metatable` to hide the metatable (or to expose the "method-table"), but what would be the benefit of constraining modders here?

We could also just make `cancel` a field in the table rather than using a metatable; there's no need to use a metatable to inherit a single field, but I suppose if we wanted to add more job methods (which ones?) this approach is better.